File: built-in plugins dsp/mb-binaural.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBinauralPlugin: PluginDefinition = { id: 'mb-binaural', slug: 'mb-binaural', name: 'MB Binaural Processor', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: 'HRTF-based binaural spatialization for headphone monitoring', author: 'Max Booster', parameters: [{ id: 'angle', name: 'Angle', type: 'float', defaultValue: 0, minValue: -180, maxValue: 180, automatable: true }, { id: 'distance', name: 'Distance', type: 'float', defaultValue: 1, minValue: 0.1, maxValue: 5, automatable: true }, { id: 'headSize', name: 'Head Size', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { angle: 0, distance: 1, headSize: 0.5, mix: 1 } };

export default MbBinauralPlugin;

File: built-in plugins dsp/mb-binaural.cpp
/**
 * MB Binaural Processor
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : HRTF-based binaural spatialization for headphone monitoring
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BINAURAL_H
#define MB_BINAURAL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBinaural : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-binaural";
    static constexpr const char* PLUGIN_NAME    = "MB Binaural Processor";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float angle = 0f;  // range [-180, 180]
    float distance = 1f;  // range [0.1, 5]
    float headSize = 0.5f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbBinaural() = default;
    ~MbBinaural() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.angle = std::clamp(params.angle, -180f, 180f);
        params.distance = std::clamp(params.distance, 0.1f, 5f);
        params.headSize = std::clamp(params.headSize, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Binaural Processor
        return input;
    }
};

#endif // MB_BINAURAL_H
